Add a content parameter to aws_s3 put operation

onitake opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Add support for uploading S3 objects from in-memory content in addition to on-disk files.

Fixes #67086
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
aws_s3

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->
Example usage:
<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below
- name: PUT operation from a rendered template
  aws_s3:
    bucket: mybucket
    object: /object.yaml
    content: "{{ lookup('template', 'templates/object.yaml.j2') }}"
    mode: put
```

srgoni committed:
Document usage of aws_s3 content parameter

srgoni committed:
Added utility function to calculate s3 etag from content instead of file

srgoni committed:
Implemented s3 put from bytes content

srgoni committed:
Moved positional parameters

srgoni committed:
Added missing imports

srgoni committed:
Replace file md5 utility call with direct calculation

srgoni committed:
pylint nitpick

tremble commented:
@onitake - FYI if you're having trouble tweaking to avoid lint issues it's possible to run the tests locally:

./bin/ansible-test sanity --docker default lib/ansible/modules/cloud/amazon/s3.py

srgoni committed:
Removed file open

onitake commented:
@tremble Sorry for abusing Shippable to do my syntax checking.
I thought I had done everything right... :grin:

srgoni committed:
Moved other use of positional param

srgoni committed:
Added potentially missing version parameter

srgoni committed:
Added unit tests for static content and template put

onitake commented:
Hmm... Looks like strings aren't autocast into byte strings.
If possible, I'd like to keep the possibility to pass a byte string into the content parameter. Is this possible at all? Should I change the `content` parameter type to string?

srgoni committed:
Changed content parameter to string and removed src requirement

srgoni committed:
Remove trailing newline from unit test

srgoni committed:
Fix unit test cleanup

srgoni committed:
Encode content string to utf-8

onitake commented:
So, I added a unit test for the content parameter and it looks like the feature's working as expected. Shippable had a timeout in a completely unrelated test, perhaps a retry is in order.

I changed the `content` parameter type to string, as I don't see a way to pass byte strings into Ansible modules. If there's a way to do this, please tell me. The string is converted to UTF-8 before sending it to S3.

What the unit tests don't cover yet is a multi-part put. This is probably crucial, as the the new `calculate_etag_content` utility function will not have full test coverage otherwise. I'll try to address this.

onitake committed:
Document implicit conversion to UTF-8

onitake commented:
Rebased to just before migration to amazon.aws (commit 6d910034a0a12f93b323e88b809c66afde6e713a).

onitake commented:
Closing in favour of https://github.com/ansible-collections/amazon.aws/pull/20
